src/prefect/engine/flow_runner.py

```diff
@@ -696,6 +696,9 @@ def determine_final_state(
             self.logger.info("Flow run SUCCESS: no reference tasks failed")
             state = Success(message="No reference tasks failed.", result=return_states)
 
+        if self.flow.terminal_state_handler:
+            return self.flow.terminal_state_handler(state)
```

Comment:
One of the major objectives for this is to be able to mutate the final state of the flow with access to additional information that is only available to the terminal state handler -- notably the final states of the tasks. The user-provided handler should be passed all this extra information. Otherwise, the user could just use a normal flow state handler to mutate final states.

Reply:
Ahhh that makes a lot more sense! I'll close this PR and re-write this